In [1]:
import pandas as pd
import matplotlib as plt
import seaborn as sns
import plotly.express as px

In [2]:
from sklearn import neighbors, preprocessing
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

In [3]:
pd.options.display.max_columns = None

Import the DataBase cleaned


In [4]:
df = pd.read_csv('DB_Valor_Prima_Anual_Cleaned.csv')

Cleaned dataset

In [5]:
df.head(3)

Tipo_poliza  Valor_Asegurado Fecha_Emisión Fecha_Inicio   Fecha_fin  \
0  Renovacion        684689000    2011-10-10   2011-11-21  2012-11-21   
1       Nuevo        965205600    2011-10-10   2011-10-05  2012-10-05   
2       Nuevo        640605600    2011-10-10   2011-10-03  2012-10-03   

   Valor_prima_Anual  Valor_asegurado_Vehiculo       Ciudad  DEPARTAMENTO  \
0           782949.0                  18800000    CARTAGENA       BOLIVAR   
1           715824.0                  24900000  BOGOTA D.C.  CUNDINAMARCA   
2           740816.0                  12600000     MEDELLIN     ANTIOQUIA   

       OCUPACION  EDAD      MARCA          REF1             REF2  \
0    EMPLEADO(A)  45.0  CHEVROLET          AVEO           FAMILY   
1    EMPLEADO(A)  41.0  CHEVROLET  AVEO EMOTION             1.6L   
2  INDEPENDIENTE  33.0    RENAULT       CLIO II  F.II EXPRESSION   

                   REF3      CLASE  Modelo_del_Vehiculo               COLOR  \
0       MT 1500CC 4P AA  AUTOMOVIL               2010.0         NEGRO EBONY   
1  MT 1600CC AA 2AB ABS  AUTOMOVIL               2009.0  BLANCO ARCO BICAPA   
2        MT 1400CC PACK  AUTOMOVIL               2002.0       BEIGE CARRARA   

      GENERO  
0  MASCULINO  
1  MASCULINO  
2   FEMENINO

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 120822 entries, 0 to 120821
Data columns (total 19 columns):
 #   Column                    Non-Null Count   Dtype  
---  ------                    --------------   -----  
 0   Tipo_poliza               120822 non-null  object 
 1   Valor_Asegurado           120822 non-null  int64  
 2   Fecha_Emisión             120822 non-null  object 
 3   Fecha_Inicio              120822 non-null  object 
 4   Fecha_fin                 120822 non-null  object 
 5   Valor_prima_Anual         120822 non-null  float64
 6   Valor_asegurado_Vehiculo  120822 non-null  int64  
 7   Ciudad                    120822 non-null  object 
 8   DEPARTAMENTO              120822 non-null  object 
 9   OCUPACION                 120822 non-null  object 
 10  EDAD                      120822 non-null  float64
 11  MARCA                     120822 non-null  object 
 12  REF1                      120822 non-null  object 
 13  REF2                      120822 non-null  o

## 1st FEATURE ENGINEERING - CREATE NEW DATA COLUMNS

I need to create new features that improve the modeling, so I decided to extract Year an Month from the datetime columns

In [7]:
df['Y_Fecha_Emisión'] = pd.DatetimeIndex(df['Fecha_Emisión']).year

df['M_Fecha_Emisión'] = pd.DatetimeIndex(df['Fecha_Emisión']).month

df['Y_Fecha_Inicio'] = pd.DatetimeIndex(df['Fecha_Inicio']).year

df['M_Fecha_Inicio'] = pd.DatetimeIndex(df['Fecha_Inicio']).month

df['Y_Fecha_fin'] = pd.DatetimeIndex(df['Fecha_fin']).year

df['M_Fecha_fin'] = pd.DatetimeIndex(df['Fecha_fin']).month

In [8]:
df.head(3)

Tipo_poliza  Valor_Asegurado Fecha_Emisión Fecha_Inicio   Fecha_fin  \
0  Renovacion        684689000    2011-10-10   2011-11-21  2012-11-21   
1       Nuevo        965205600    2011-10-10   2011-10-05  2012-10-05   
2       Nuevo        640605600    2011-10-10   2011-10-03  2012-10-03   

   Valor_prima_Anual  Valor_asegurado_Vehiculo       Ciudad  DEPARTAMENTO  \
0           782949.0                  18800000    CARTAGENA       BOLIVAR   
1           715824.0                  24900000  BOGOTA D.C.  CUNDINAMARCA   
2           740816.0                  12600000     MEDELLIN     ANTIOQUIA   

       OCUPACION  EDAD      MARCA          REF1             REF2  \
0    EMPLEADO(A)  45.0  CHEVROLET          AVEO           FAMILY   
1    EMPLEADO(A)  41.0  CHEVROLET  AVEO EMOTION             1.6L   
2  INDEPENDIENTE  33.0    RENAULT       CLIO II  F.II EXPRESSION   

                   REF3      CLASE  Modelo_del_Vehiculo               COLOR  \
0       MT 1500CC 4P AA  AUTOMOVIL               2010.0         NEGRO EBONY   
1  MT 1600CC AA 2AB ABS  AUTOMOVIL               2009.0  BLANCO ARCO BICAPA   
2        MT 1400CC PACK  AUTOMOVIL               2002.0       BEIGE CARRARA   

      GENERO  Y_Fecha_Emisión  M_Fecha_Emisión  Y_Fecha_Inicio  \
0  MASCULINO             2011               10            2011   
1  MASCULINO             2011               10            2011   
2   FEMENINO             2011               10            2011   

   M_Fecha_Inicio  Y_Fecha_fin  M_Fecha_fin  
0              11         2012           11  
1              10         2012           10  
2              10         2012           10

# 1. TRAIN TEST SPLIT

Target variable = Valor_prima_Anual

In [9]:
X = df.drop(['Valor_prima_Anual', 'REF1', 'REF2', 'REF3', 'MARCA', 'Fecha_Emisión', 'Fecha_Inicio', 'Fecha_fin'], axis=1)
Y = df['Valor_prima_Anual']

X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=0.2, random_state=42)

In [10]:
X_train.head(2)

Tipo_poliza  Valor_Asegurado  Valor_asegurado_Vehiculo Ciudad  \
108407       Nuevo       3089479013                  34900000   CALI   
26677   Renovacion        664977882                  24300000   CALI   

       DEPARTAMENTO    OCUPACION  EDAD      CLASE  Modelo_del_Vehiculo  \
108407        VALLE  EMPLEADO(A)  30.0  AUTOMOVIL               2014.0   
26677         VALLE  EMPLEADO(A)  56.0  AUTOMOVIL               2006.0   

         COLOR    GENERO  Y_Fecha_Emisión  M_Fecha_Emisión  Y_Fecha_Inicio  \
108407  PLATA   FEMENINO             2013               11            2013   
26677    ARENA  FEMENINO             2012                9            2012   

        M_Fecha_Inicio  Y_Fecha_fin  M_Fecha_fin  
108407              11         2014           11  
26677                9         2013            9

# PIPELINE TO TRANSFORM VARIABLES

In [11]:
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler, OneHotEncoder, PolynomialFeatures

numeric_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='median')),
    ('poly', PolynomialFeatures(degree=2)),
    ('scaler', StandardScaler())])
categorical_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='constant', fill_value='missing')),
    ('onehot', OneHotEncoder(handle_unknown='ignore'))])

In [12]:
numeric_features = X_train.select_dtypes(include=['int64', 'float64']).columns
categorical_features = X_train.select_dtypes(include=['object', 'category']).columns

from sklearn.compose import ColumnTransformer

preprocessor = ColumnTransformer(
    transformers=[
        ('num', numeric_transformer, numeric_features),
        ('cat', categorical_transformer, categorical_features)])

# 1. MODEL CREATION

## RUN #1

X = df.drop(['Valor_prima_Anual', 'REF1', 'REF2', 'REF3', 'MARCA', 'Fecha_Emisión', 'Fecha_Inicio', 'Fecha_fin'], axis=1) <br>
Y = df['Valor_prima_Anual']

In [13]:
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

from sklearn.linear_model import LinearRegression
from sklearn.neighbors import KNeighborsRegressor
from sklearn.svm import SVR, LinearSVR, NuSVR
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor, AdaBoostRegressor, GradientBoostingRegressor

from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
from sklearn.discriminant_analysis import QuadraticDiscriminantAnalysis

from xgboost import XGBRegressor
from lightgbm import LGBMRegressor

regressors = [
    LinearRegression(),
    # KNeighborsRegressor(3),
    # SVR(kernel="rbf", C=0.025),
    # DecisionTreeRegressor(),
    # RandomForestRegressor(),
    # AdaBoostRegressor(),
    # GradientBoostingRegressor(),
    # XGBRegressor(),
    LGBMRegressor()
    ]
for regressor in regressors:
    pipe = Pipeline(steps=[('preprocessor', preprocessor),
                      ('regressor', regressor)])
    pipe.fit(X_train, y_train)   
    y_pred = pipe.predict(X_test)
    print(regressor)
    print("model MAE : " , mean_absolute_error(y_test, y_pred))
    print("model MSE : " ,  mean_squared_error(y_test, y_pred))
    print("model R2  : " ,  r2_score(y_test, y_pred))

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)
model MAE :  163042.47129452188
model MSE :  67787253411.01022
model R2  :  0.7287738559441437
LGBMRegressor(boosting_type='gbdt', class_weight=None, colsample_bytree=1.0,
              importance_type='split', learning_rate=0.1, max_depth=-1,
              min_child_samples=20, min_child_weight=0.001, min_split_gain=0.0,
              n_estimators=100, n_jobs=-1, num_leaves=31, objective=None,
              random_state=None, reg_alpha=0.0, reg_lambda=0.0, silent=True,
              subsample=1.0, subsample_for_bin=200000, subsample_freq=0)
model MAE :  94282.25687616429
model MSE :  26698095445.86297
model R2  :  0.893177240306354


## RUN #2

## 2nd FEATURE ENGINEERING - CREATE NEW DATA COLUMNS

In [14]:
df['MT_AT_TP_REF3'] = df['REF3'].astype(str).str[0:2]
df['MT_AT_TP_REF3']

0         MT
1         MT
2         MT
3         MT
4         MT
          ..
120817    MT
120818    MT
120819    AT
120820    MT
120821    MT
Name: MT_AT_TP_REF3, Length: 120822, dtype: object

In [15]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 120822 entries, 0 to 120821
Data columns (total 26 columns):
 #   Column                    Non-Null Count   Dtype  
---  ------                    --------------   -----  
 0   Tipo_poliza               120822 non-null  object 
 1   Valor_Asegurado           120822 non-null  int64  
 2   Fecha_Emisión             120822 non-null  object 
 3   Fecha_Inicio              120822 non-null  object 
 4   Fecha_fin                 120822 non-null  object 
 5   Valor_prima_Anual         120822 non-null  float64
 6   Valor_asegurado_Vehiculo  120822 non-null  int64  
 7   Ciudad                    120822 non-null  object 
 8   DEPARTAMENTO              120822 non-null  object 
 9   OCUPACION                 120822 non-null  object 
 10  EDAD                      120822 non-null  float64
 11  MARCA                     120822 non-null  object 
 12  REF1                      120822 non-null  object 
 13  REF2                      120822 non-null  o

In [16]:
X = df.drop(['Valor_prima_Anual', 'REF1', 'REF2', 'REF3', 'MARCA', 'Fecha_Emisión', 'Fecha_Inicio', 'Fecha_fin'], axis=1)
Y = df['Valor_prima_Anual']

X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=0.2, random_state=42)

In [17]:
X_train.head(2)

Tipo_poliza  Valor_Asegurado  Valor_asegurado_Vehiculo Ciudad  \
108407       Nuevo       3089479013                  34900000   CALI   
26677   Renovacion        664977882                  24300000   CALI   

       DEPARTAMENTO    OCUPACION  EDAD      CLASE  Modelo_del_Vehiculo  \
108407        VALLE  EMPLEADO(A)  30.0  AUTOMOVIL               2014.0   
26677         VALLE  EMPLEADO(A)  56.0  AUTOMOVIL               2006.0   

         COLOR    GENERO  Y_Fecha_Emisión  M_Fecha_Emisión  Y_Fecha_Inicio  \
108407  PLATA   FEMENINO             2013               11            2013   
26677    ARENA  FEMENINO             2012                9            2012   

        M_Fecha_Inicio  Y_Fecha_fin  M_Fecha_fin MT_AT_TP_REF3  
108407              11         2014           11            MT  
26677                9         2013            9            MT

In [18]:
numeric_features = X_train.select_dtypes(include=['int64', 'float64']).columns
categorical_features = X_train.select_dtypes(include=['object', 'category']).columns

from sklearn.compose import ColumnTransformer

preprocessor = ColumnTransformer(
    transformers=[
        ('num', numeric_transformer, numeric_features),
        ('cat', categorical_transformer, categorical_features)])

In [19]:
numeric_features

Index(['Valor_Asegurado', 'Valor_asegurado_Vehiculo', 'EDAD',
       'Modelo_del_Vehiculo', 'Y_Fecha_Emisión', 'M_Fecha_Emisión',
       'Y_Fecha_Inicio', 'M_Fecha_Inicio', 'Y_Fecha_fin', 'M_Fecha_fin'],
      dtype='object')

In [20]:
categorical_features

Index(['Tipo_poliza', 'Ciudad', 'DEPARTAMENTO', 'OCUPACION', 'CLASE', 'COLOR',
       'GENERO', 'MT_AT_TP_REF3'],
      dtype='object')

In [21]:
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

from sklearn.linear_model import LinearRegression
from sklearn.neighbors import KNeighborsRegressor
from sklearn.svm import SVR, LinearSVR, NuSVR
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor, AdaBoostRegressor, GradientBoostingRegressor

from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
from sklearn.discriminant_analysis import QuadraticDiscriminantAnalysis

from xgboost import XGBRegressor
from lightgbm import LGBMRegressor

regressors = [
    LinearRegression(),
    # KNeighborsRegressor(3),
    # SVR(kernel="rbf", C=0.025),
    # DecisionTreeRegressor(),
    # RandomForestRegressor(),
    # AdaBoostRegressor(),
    # GradientBoostingRegressor(),
    # XGBRegressor(),
    LGBMRegressor()
    ]
for regressor in regressors:
    pipe = Pipeline(steps=[('preprocessor', preprocessor),
                      ('regressor', regressor)])
    pipe.fit(X_train, y_train)   
    y_pred = pipe.predict(X_test)
    print(regressor)
    print("model MAE : " , mean_absolute_error(y_test, y_pred))
    print("model MSE : " ,  mean_squared_error(y_test, y_pred))
    print("model R2  : " ,  r2_score(y_test, y_pred))

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)
model MAE :  162861.50596567488
model MSE :  67829410827.43192
model R2  :  0.7286051783104581
LGBMRegressor(boosting_type='gbdt', class_weight=None, colsample_bytree=1.0,
              importance_type='split', learning_rate=0.1, max_depth=-1,
              min_child_samples=20, min_child_weight=0.001, min_split_gain=0.0,
              n_estimators=100, n_jobs=-1, num_leaves=31, objective=None,
              random_state=None, reg_alpha=0.0, reg_lambda=0.0, silent=True,
              subsample=1.0, subsample_for_bin=200000, subsample_freq=0)
model MAE :  93795.85958618265
model MSE :  26298012667.089348
model R2  :  0.8947780266478802


## RUN #3

## 3rd FEATURE ENGINEERING - CREATE NEW DATA COLUMNS

In [22]:
df['CC_REF3'] = df['REF3'].str.extract('(\d+)')
df['CC_REF3'].unique()
df['CC_REF3'] = df['CC_REF3'].astype('int64')

In [23]:
X = df.drop(['Valor_prima_Anual', 'REF1', 'REF2', 'REF3', 'MARCA', 'Fecha_Emisión', 'Fecha_Inicio', 'Fecha_fin'], axis=1)
Y = df['Valor_prima_Anual']

X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=0.2, random_state=42)

In [24]:
X_train.head(2)

Tipo_poliza  Valor_Asegurado  Valor_asegurado_Vehiculo Ciudad  \
108407       Nuevo       3089479013                  34900000   CALI   
26677   Renovacion        664977882                  24300000   CALI   

       DEPARTAMENTO    OCUPACION  EDAD      CLASE  Modelo_del_Vehiculo  \
108407        VALLE  EMPLEADO(A)  30.0  AUTOMOVIL               2014.0   
26677         VALLE  EMPLEADO(A)  56.0  AUTOMOVIL               2006.0   

         COLOR    GENERO  Y_Fecha_Emisión  M_Fecha_Emisión  Y_Fecha_Inicio  \
108407  PLATA   FEMENINO             2013               11            2013   
26677    ARENA  FEMENINO             2012                9            2012   

        M_Fecha_Inicio  Y_Fecha_fin  M_Fecha_fin MT_AT_TP_REF3  CC_REF3  
108407              11         2014           11            MT     1200  
26677                9         2013            9            MT     1600

In [25]:
numeric_features = X_train.select_dtypes(include=['int64', 'float64']).columns
categorical_features = X_train.select_dtypes(include=['object', 'category']).columns

from sklearn.compose import ColumnTransformer

preprocessor = ColumnTransformer(
    transformers=[
        ('num', numeric_transformer, numeric_features),
        ('cat', categorical_transformer, categorical_features)])

In [26]:
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

from sklearn.linear_model import LinearRegression
from sklearn.neighbors import KNeighborsRegressor
from sklearn.svm import SVR, LinearSVR, NuSVR
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor, AdaBoostRegressor, GradientBoostingRegressor

from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
from sklearn.discriminant_analysis import QuadraticDiscriminantAnalysis

from xgboost import XGBRegressor
from lightgbm import LGBMRegressor

regressors = [
    LinearRegression(),
    # KNeighborsRegressor(3),
    # SVR(kernel="rbf", C=0.025),
    # DecisionTreeRegressor(),
    # RandomForestRegressor(),
    # AdaBoostRegressor(),
    # GradientBoostingRegressor(),
    # XGBRegressor(),
    LGBMRegressor()
    ]
for regressor in regressors:
    pipe = Pipeline(steps=[('preprocessor', preprocessor),
                      ('regressor', regressor)])
    pipe.fit(X_train, y_train)   
    y_pred = pipe.predict(X_test)
    print(regressor)
    print("model MAE : " , mean_absolute_error(y_test, y_pred))
    print("model MSE : " ,  mean_squared_error(y_test, y_pred))
    print("model R2  : " ,  r2_score(y_test, y_pred))

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)
model MAE :  162922.82767212923
model MSE :  67667245782.69293
model R2  :  0.7292540230057656
LGBMRegressor(boosting_type='gbdt', class_weight=None, colsample_bytree=1.0,
              importance_type='split', learning_rate=0.1, max_depth=-1,
              min_child_samples=20, min_child_weight=0.001, min_split_gain=0.0,
              n_estimators=100, n_jobs=-1, num_leaves=31, objective=None,
              random_state=None, reg_alpha=0.0, reg_lambda=0.0, silent=True,
              subsample=1.0, subsample_for_bin=200000, subsample_freq=0)
model MAE :  94152.81826766733
model MSE :  26044571444.09245
model R2  :  0.8957920799130489


## RUN #4

## 3rd FEATURE ENGINEERING - ADDING NEW DATA COLUMNS "MARCA"

In [27]:
df.head(2)

Tipo_poliza  Valor_Asegurado Fecha_Emisión Fecha_Inicio   Fecha_fin  \
0  Renovacion        684689000    2011-10-10   2011-11-21  2012-11-21   
1       Nuevo        965205600    2011-10-10   2011-10-05  2012-10-05   

   Valor_prima_Anual  Valor_asegurado_Vehiculo       Ciudad  DEPARTAMENTO  \
0           782949.0                  18800000    CARTAGENA       BOLIVAR   
1           715824.0                  24900000  BOGOTA D.C.  CUNDINAMARCA   

     OCUPACION  EDAD      MARCA          REF1    REF2                  REF3  \
0  EMPLEADO(A)  45.0  CHEVROLET          AVEO  FAMILY       MT 1500CC 4P AA   
1  EMPLEADO(A)  41.0  CHEVROLET  AVEO EMOTION    1.6L  MT 1600CC AA 2AB ABS   

       CLASE  Modelo_del_Vehiculo               COLOR     GENERO  \
0  AUTOMOVIL               2010.0         NEGRO EBONY  MASCULINO   
1  AUTOMOVIL               2009.0  BLANCO ARCO BICAPA  MASCULINO   

   Y_Fecha_Emisión  M_Fecha_Emisión  Y_Fecha_Inicio  M_Fecha_Inicio  \
0             2011               10            2011              11   
1             2011               10            2011              10   

   Y_Fecha_fin  M_Fecha_fin MT_AT_TP_REF3  CC_REF3  
0         2012           11            MT     1500  
1         2012           10            MT     1600

In [28]:
X = df.drop(['Valor_prima_Anual', 'REF1', 'REF2', 'REF3', 'Fecha_Emisión', 'Fecha_Inicio', 'Fecha_fin'], axis=1)
Y = df['Valor_prima_Anual']

X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=0.2, random_state=42)

In [29]:
X_train.head(2)

Tipo_poliza  Valor_Asegurado  Valor_asegurado_Vehiculo Ciudad  \
108407       Nuevo       3089479013                  34900000   CALI   
26677   Renovacion        664977882                  24300000   CALI   

       DEPARTAMENTO    OCUPACION  EDAD  MARCA      CLASE  Modelo_del_Vehiculo  \
108407        VALLE  EMPLEADO(A)  30.0    KIA  AUTOMOVIL               2014.0   
26677         VALLE  EMPLEADO(A)  56.0  MAZDA  AUTOMOVIL               2006.0   

         COLOR    GENERO  Y_Fecha_Emisión  M_Fecha_Emisión  Y_Fecha_Inicio  \
108407  PLATA   FEMENINO             2013               11            2013   
26677    ARENA  FEMENINO             2012                9            2012   

        M_Fecha_Inicio  Y_Fecha_fin  M_Fecha_fin MT_AT_TP_REF3  CC_REF3  
108407              11         2014           11            MT     1200  
26677                9         2013            9            MT     1600

In [30]:
numeric_features = X_train.select_dtypes(include=['int64', 'float64']).columns
categorical_features = X_train.select_dtypes(include=['object', 'category']).columns

from sklearn.compose import ColumnTransformer

preprocessor = ColumnTransformer(
    transformers=[
        ('num', numeric_transformer, numeric_features),
        ('cat', categorical_transformer, categorical_features)])

In [31]:
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

from sklearn.linear_model import LinearRegression
from sklearn.neighbors import KNeighborsRegressor
from sklearn.svm import SVR, LinearSVR, NuSVR
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor, AdaBoostRegressor, GradientBoostingRegressor

from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
from sklearn.discriminant_analysis import QuadraticDiscriminantAnalysis

from xgboost import XGBRegressor
from lightgbm import LGBMRegressor

regressors = [
    LinearRegression(),
    # KNeighborsRegressor(3),
    # SVR(kernel="rbf", C=0.025),
    # DecisionTreeRegressor(),
    # RandomForestRegressor(),
    # AdaBoostRegressor(),
    # GradientBoostingRegressor(),
    # XGBRegressor(),
    LGBMRegressor()
    ]
for regressor in regressors:
    pipe = Pipeline(steps=[('preprocessor', preprocessor),
                      ('regressor', regressor)])
    pipe.fit(X_train, y_train)   
    y_pred = pipe.predict(X_test)
    print(regressor)
    print("model MAE : " , mean_absolute_error(y_test, y_pred))
    print("model MSE : " ,  mean_squared_error(y_test, y_pred))
    print("model R2  : " ,  r2_score(y_test, y_pred))

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)
model MAE :  161914.87817033587
model MSE :  67039256218.91058
model R2  :  0.7317666958066427
LGBMRegressor(boosting_type='gbdt', class_weight=None, colsample_bytree=1.0,
              importance_type='split', learning_rate=0.1, max_depth=-1,
              min_child_samples=20, min_child_weight=0.001, min_split_gain=0.0,
              n_estimators=100, n_jobs=-1, num_leaves=31, objective=None,
              random_state=None, reg_alpha=0.0, reg_lambda=0.0, silent=True,
              subsample=1.0, subsample_for_bin=200000, subsample_freq=0)
model MAE :  92779.1713959528
model MSE :  24875340462.45637
model R2  :  0.9004703342263928


# CONCLUSION OF MODELING AND TESTING

1. The first thing to do to improve the results of a prediction model is try to figure out the high feature engineering, after this try to do a hyperruning to your model. 
2. using others models the time to train the model was so long (2 hrs), due to the amount of unique values of certain values
3. Even doing some transformations on those columns, the time was so long, I decided to work only with LR and LGBMR
4. Doing some feature engineering on the columns like REF, FECHAS and others I could increase the R2 from 0.8 to 0.9

# BEST MODEL = LGBMRegressor
model MAE :  92779.1713959528 <br>
model MSE :  24875340462.45637 <br>
model R2  :  0.9004703342263928

In [32]:
LGBMRegressor()

LGBMRegressor(boosting_type='gbdt', class_weight=None, colsample_bytree=1.0,
              importance_type='split', learning_rate=0.1, max_depth=-1,
              min_child_samples=20, min_child_weight=0.001, min_split_gain=0.0,
              n_estimators=100, n_jobs=-1, num_leaves=31, objective=None,
              random_state=None, reg_alpha=0.0, reg_lambda=0.0, silent=True,
              subsample=1.0, subsample_for_bin=200000, subsample_freq=0)

# BONUS = Let´s add a new feture "ANTIGUEDAD", and let´s see how is its behaviour

# RUN #5

In [33]:
df2 = pd.read_csv('DB_Valor_Prima_Anual_Cleaned_antiguedadCol.csv')

In [34]:
df2['Y_Fecha_Emisión'] = pd.DatetimeIndex(df2['Fecha_Emisión']).year

df2['M_Fecha_Emisión'] = pd.DatetimeIndex(df2['Fecha_Emisión']).month

df2['Y_Fecha_Inicio'] = pd.DatetimeIndex(df2['Fecha_Inicio']).year

df2['M_Fecha_Inicio'] = pd.DatetimeIndex(df2['Fecha_Inicio']).month

df2['Y_Fecha_fin'] = pd.DatetimeIndex(df2['Fecha_fin']).year

df2['M_Fecha_fin'] = pd.DatetimeIndex(df2['Fecha_fin']).month

In [35]:
df2['MT_AT_TP_REF3'] = df2['REF3'].astype(str).str[0:2]

In [36]:
df2['CC_REF3'] = df2['REF3'].str.extract('(\d+)')
df2['CC_REF3'].unique()
df2['CC_REF3'] = df2['CC_REF3'].astype('int64')

In [37]:
X = df2.drop(['Valor_prima_Anual', 'REF1', 'REF2', 'REF3', 'Fecha_Emisión', 'Fecha_Inicio', 'Fecha_fin'], axis=1)
Y = df2['Valor_prima_Anual']

X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=0.2, random_state=42)

In [38]:
X_train.head(3)

Tipo_poliza  Valor_Asegurado  Valor_asegurado_Vehiculo    Ciudad  \
108407       Nuevo       3089479013                  34900000      CALI   
26677   Renovacion        664977882                  24300000      CALI   
41802        Nuevo       3067781882                  24100000  MEDELLIN   

       DEPARTAMENTO    OCUPACION  EDAD    MARCA      CLASE  \
108407        VALLE  EMPLEADO(A)  30.0      KIA  AUTOMOVIL   
26677         VALLE  EMPLEADO(A)  56.0    MAZDA  AUTOMOVIL   
41802     ANTIOQUIA  EMPLEADO(A)  36.0  RENAULT  AUTOMOVIL   

        Modelo_del_Vehiculo   COLOR    GENERO  ANTIGUEDAD  Y_Fecha_Emisión  \
108407               2014.0  PLATA   FEMENINO         365             2013   
26677                2006.0   ARENA  FEMENINO         365             2012   
41802                2012.0    GRIS  FEMENINO         730             2012   

        M_Fecha_Emisión  Y_Fecha_Inicio  M_Fecha_Inicio  Y_Fecha_fin  \
108407               11            2013              11         2014   
26677                 9            2012               9         2013   
41802                11            2012              11         2014   

        M_Fecha_fin MT_AT_TP_REF3  CC_REF3  
108407           11            MT     1200  
26677             9            MT     1600  
41802            11            MT     1200

In [39]:
numeric_features = X_train.select_dtypes(include=['int64', 'float64']).columns
categorical_features = X_train.select_dtypes(include=['object', 'category']).columns

from sklearn.compose import ColumnTransformer

preprocessor = ColumnTransformer(
    transformers=[
        ('num', numeric_transformer, numeric_features),
        ('cat', categorical_transformer, categorical_features)])

In [40]:
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

from sklearn.linear_model import LinearRegression
from sklearn.neighbors import KNeighborsRegressor
from sklearn.svm import SVR, LinearSVR, NuSVR
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor, AdaBoostRegressor, GradientBoostingRegressor

from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
from sklearn.discriminant_analysis import QuadraticDiscriminantAnalysis

from xgboost import XGBRegressor
from lightgbm import LGBMRegressor

regressors = [
    LinearRegression(),
    # KNeighborsRegressor(3),
    # SVR(kernel="rbf", C=0.025),
    # DecisionTreeRegressor(),
    # RandomForestRegressor(),
    # AdaBoostRegressor(),
    # GradientBoostingRegressor(),
    # XGBRegressor(),
    LGBMRegressor()
    ]
for regressor in regressors:
    pipe = Pipeline(steps=[('preprocessor', preprocessor),
                      ('regressor', regressor)])
    pipe.fit(X_train, y_train)   
    y_pred = pipe.predict(X_test)
    print(regressor)
    print("model MAE : " , mean_absolute_error(y_test, y_pred))
    print("model MSE : " ,  mean_squared_error(y_test, y_pred))
    print("model R2  : " ,  r2_score(y_test, y_pred))

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)
model MAE :  159263.53713051035
model MSE :  65395912023.48509
model R2  :  0.738341942435671
LGBMRegressor(boosting_type='gbdt', class_weight=None, colsample_bytree=1.0,
              importance_type='split', learning_rate=0.1, max_depth=-1,
              min_child_samples=20, min_child_weight=0.001, min_split_gain=0.0,
              n_estimators=100, n_jobs=-1, num_leaves=31, objective=None,
              random_state=None, reg_alpha=0.0, reg_lambda=0.0, silent=True,
              subsample=1.0, subsample_for_bin=200000, subsample_freq=0)
model MAE :  92081.95219005925
model MSE :  24838149171.01345
model R2  :  0.9006191417095543


it increase but a little 

# FEATURE IMPORTANCE

## With feature importance we could see the variables that we need to concern about, the most important variables to the machine.

In [41]:
import lightgbm

In [42]:

feature= df.drop(['Valor_prima_Anual', 'REF1', 'REF2', 'REF3', 'Fecha_Emisión', 'Fecha_Inicio', 'Fecha_fin'], axis=1)
target= df[['Valor_prima_Anual']]
feature_train, feature_test, target_train, target_test= train_test_split(feature, target, test_size=0.12)

In [43]:
print('total feature training features: ', len(feature_train))
print('total feature testing features: ', len(feature_test))
print('total target training features: ', len(target_train))
print('total target testing features: ', len(target_test))

total feature training features:  106323
total feature testing features:  14499
total target training features:  106323
total target testing features:  14499


In [44]:
feature_train.shape , target_train.shape



((106323, 20), (106323, 1))

In [45]:
train_data = lightgbm.Dataset(feature_train, label=target_train, categorical_feature=categorical_features)
test_data = lightgbm.Dataset(feature_test, label=target_test)

In [46]:
#basic parameter:
parameters = {
    'application': 'binary',
    'objective': 'binary',
    'metric': 'auc',
    'is_unbalance': 'true',
    'boosting': 'gbdt',
    'num_leaves': 31,
    'feature_fraction': 0.5,
    'bagging_fraction': 0.5,
    'bagging_freq': 20,
    'learning_rate': 0.05,
    'verbose': 0
}

In [47]:
model = lightgbm.train(parameters,
                       train_data,
                       valid_sets=test_data,
                       num_boost_round=5000,
                       early_stopping_rounds=100)

ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()

# HYPERPARAMETER TUNNING

In [48]:
X = df.drop(['Valor_prima_Anual', 'REF1', 'REF2', 'REF3', 'Fecha_Emisión', 'Fecha_Inicio', 'Fecha_fin'], axis=1)
Y = df['Valor_prima_Anual']

X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=0.2, random_state=42)

In [49]:
preprocessor = ColumnTransformer(
    transformers=[
        ('num', numeric_transformer, numeric_features),
        ('cat', categorical_transformer, categorical_features)])

In [50]:
X_train.head(2)

Tipo_poliza  Valor_Asegurado  Valor_asegurado_Vehiculo Ciudad  \
108407       Nuevo       3089479013                  34900000   CALI   
26677   Renovacion        664977882                  24300000   CALI   

       DEPARTAMENTO    OCUPACION  EDAD  MARCA      CLASE  Modelo_del_Vehiculo  \
108407        VALLE  EMPLEADO(A)  30.0    KIA  AUTOMOVIL               2014.0   
26677         VALLE  EMPLEADO(A)  56.0  MAZDA  AUTOMOVIL               2006.0   

         COLOR    GENERO  Y_Fecha_Emisión  M_Fecha_Emisión  Y_Fecha_Inicio  \
108407  PLATA   FEMENINO             2013               11            2013   
26677    ARENA  FEMENINO             2012                9            2012   

        M_Fecha_Inicio  Y_Fecha_fin  M_Fecha_fin MT_AT_TP_REF3  CC_REF3  
108407              11         2014           11            MT     1200  
26677                9         2013            9            MT     1600

In [51]:
from sklearn.model_selection import KFold, cross_val_score, GridSearchCV

In [52]:
LGBMR = LGBMRegressor()

LGBMR_param_grid = {'objective':['regression','dart','goss','rf'],
                    'num_leaves':[7],
                    'learning_rate':[0.01], 
                    'n_estimators': [3300],
                    'max_depth':[4], 
                    'max_bin': [65],
                    'bagging_fraction':[0.6],
                    'bagging_freq':[9], 
                    'feature_fraction':[0.1],
                    'feature_fraction_seed':[1],
                    'bagging_seed':[14],
                    'min_data_in_leaf':[5], 
                    'min_sum_hessian_in_leaf':[5],
                    'colsample_bytree':[0],
                    'reg_alpha':[0.2],
                    'reg_lambda':[0.1]}

gsLGBMR = GridSearchCV(LGBMR, 
                      param_grid = LGBMR_param_grid, 
                      cv = 10, 
                      scoring = "neg_mean_squared_error", 
                      n_jobs = -1, 
                      verbose = 1)

gsLGBMR.fit(X_train,y_train)

LGBMR_best = gsLGBMR.best_estimator_

gsLGBMR.best_score_

Fitting 10 folds for each of 4 candidates, totalling 40 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=-1)]: Done  40 out of  40 | elapsed:   12.6s finished


ValueError: DataFrame.dtypes for data must be int, float or bool.
Did not expect the data types in the following fields: Tipo_poliza, Ciudad, DEPARTAMENTO, OCUPACION, MARCA, CLASE, COLOR, GENERO, MT_AT_TP_REF3

# CONCLUSION HYPERTUNING AND FEATURING IMPORTANCES

1. Its necesary to do an optimal hypertuning, but due to the lack of time I couldnt finish it.
2. Look for the feature importances allow us to focus on the activities that boost the business
3. Would be a great idea to develop a PowerBI or TABLEAU project.

# HYPERTUNNING WITH PIPELINE

In [53]:
from lightgbm import LGBMRegressor

LGBMR = Pipeline(steps=[('preprocessor', preprocessor),
                      ('classifier', LGBMRegressor())])

In [54]:
param_grid = { 
    'classifier__n_estimators': [600,900],
    'classifier__max_depth' : [9,11]}
from sklearn.model_selection import GridSearchCV
CV = GridSearchCV(LGBMR, param_grid, n_jobs= 1)
                  
CV.fit(X_train, y_train)  
print(CV.best_params_)    
print(CV.best_score_)

C:\Users\andre\anaconda3\lib\site-packages\sklearn\model_selection\_validation.py:536: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: 
ValueError: A given column is not a column of the dataframe

  FitFailedWarning)


ValueError: A given column is not a column of the dataframe

# HYPERPARAMETER TUNNING WITH OPTUNA

In [59]:
categorical_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='constant', fill_value='missing')),
    ('onehot', OneHotEncoder(handle_unknown='ignore'))])

In [60]:
numeric_features = X_train.select_dtypes(include=['int64', 'float64']).columns
categorical_features = X_train.select_dtypes(include=['object', 'category']).columns

from sklearn.compose import ColumnTransformer

preprocessor = ColumnTransformer(
    transformers=[
        ('num', numeric_transformer, numeric_features),
        ('cat', categorical_transformer, categorical_features)])

In [64]:
categorical_features

Index(['Tipo_poliza', 'Ciudad', 'DEPARTAMENTO', 'OCUPACION', 'MARCA', 'CLASE',
       'COLOR', 'GENERO', 'MT_AT_TP_REF3'],
      dtype='object')

In [90]:
X_train_dummies = pd.get_dummies(df, columns = categorical_features)


In [91]:
X_train_dummies.head(2)

Valor_Asegurado Fecha_Emisión Fecha_Inicio   Fecha_fin  Valor_prima_Anual  \
0        684689000    2011-10-10   2011-11-21  2012-11-21           782949.0   
1        965205600    2011-10-10   2011-10-05  2012-10-05           715824.0   

   Valor_asegurado_Vehiculo  EDAD          REF1    REF2                  REF3  \
0                  18800000  45.0          AVEO  FAMILY       MT 1500CC 4P AA   
1                  24900000  41.0  AVEO EMOTION    1.6L  MT 1600CC AA 2AB ABS   

   Modelo_del_Vehiculo  Y_Fecha_Emisión  M_Fecha_Emisión  Y_Fecha_Inicio  \
0               2010.0             2011               10            2011   
1               2009.0             2011               10            2011   

   M_Fecha_Inicio  Y_Fecha_fin  M_Fecha_fin  CC_REF3  \
0              11         2012           11     1500   
1              10         2012           10     1600   

   Tipo_poliza_Modificación  Tipo_poliza_Nuevo  Tipo_poliza_Renovacion  \
0                         0                  0                       1   
1                         0                  1                       0   

   Ciudad_ABEJORRAL  Ciudad_ABREGO  Ciudad_ACACIAS  Ciudad_ACEVEDO  \
0                 0              0               0               0   
1                 0              0               0               0   

   Ciudad_AGUA DE DIOS  Ciudad_AGUACHICA  Ciudad_AGUADAS  Ciudad_AGUAZUL  \
0                    0                 0               0               0   
1                    0                 0               0               0   

   Ciudad_AGUSTIN CODAZZI  Ciudad_AIPE  Ciudad_ALBANIA  Ciudad_ALBANIA (CAQ.)  \
0                       0            0               0                      0   
1                       0            0               0                      0   

   Ciudad_ALBANIA (GUAJIRA)  Ciudad_ALCALA  Ciudad_ALGECIRAS  Ciudad_ALMAGUER  \
0                         0              0                 0                0   
1                         0              0                 0                0   

   Ciudad_ALPUJARRA  Ciudad_ALTAMIRA  Ciudad_AMAGA  Ciudad_AMALFI  \
0                 0                0             0              0   
1                 0                0             0              0   

   Ciudad_AMBALEMA  Ciudad_ANAPOIMA  Ciudad_ANDALUCIA  Ciudad_ANDES  \
0                0                0                 0             0   
1                0                0                 0             0   

   Ciudad_ANORI  Ciudad_ANSERMA  Ciudad_ANSERMANUEVO  Ciudad_ANTIOQUIA  \
0             0               0                    0                 0   
1             0               0                    0                 0   

   Ciudad_ANZOATEGUI  Ciudad_APARTADO  Ciudad_APIA  Ciudad_AQUITANIA  \
0                  0                0            0                 0   
1                  0                0            0                 0   

   Ciudad_ARACATACA  Ciudad_ARANZAZU  Ciudad_ARATOCA  Ciudad_ARAUCA  \
0                 0                0               0              0   
1                 0                0               0              0   

   Ciudad_ARAUQUITA  Ciudad_ARBELAEZ  Ciudad_ARBOLEDAS  Ciudad_ARBOLETES  \
0                 0                0                 0                 0   
1                 0                0                 0                 0   

   Ciudad_ARGELIA (VALLE)  Ciudad_ARJONA  Ciudad_ARMENIA (ANT.)  \
0                       0              0                      0   
1                       0              0                      0   

   Ciudad_ARMENIA (QUINDIO)  Ciudad_ASTREA  Ciudad_AÑAPEL  \
0                         0              0              0   
1                         0              0              0   

   Ciudad_BALBOA (R/LDA)  Ciudad_BARANOA  Ciudad_BARAYA  Ciudad_BARBACOAS  \
0                      0               0              0                 0   
1                      0               0              0                 0   

   Ciudad_BARBOSA (ANT.)  Ciudad_BARBOSA (S/DE

In [80]:
X = X_train_dummies.drop(['Valor_prima_Anual', 'REF1', 'REF2', 'REF3', 'Fecha_Emisión', 'Fecha_Inicio', 'Fecha_fin'], axis=1)
Y = df2['Valor_prima_Anual']

In [81]:
X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=0.2, random_state=42)

In [79]:
from lightgbm import LGBMRegressor


In [92]:
def objective(trial,data=X_train,target=y_train):
    
    train_x, test_x, train_y, test_y = train_test_split(data, target, test_size=0.2,random_state=42)
    train_x = X
    train_y = Y 
    test_x = X_test 
    test_y = y_test
    param = {
        'metric': 'rmse', 
        'random_state': 48,
        'n_estimators': 20000,
        'reg_alpha': trial.suggest_loguniform('reg_alpha', 1e-3, 10.0),
        'reg_lambda': trial.suggest_loguniform('reg_lambda', 1e-3, 10.0),
        'colsample_bytree': trial.suggest_categorical('colsample_bytree', [0.3,0.4,0.5,0.6,0.7,0.8,0.9, 1.0]),
        'subsample': trial.suggest_categorical('subsample', [0.4,0.5,0.6,0.7,0.8,1.0]),
        'learning_rate': trial.suggest_categorical('learning_rate', [0.006,0.008,0.01,0.014,0.017,0.02]),
        'max_depth': trial.suggest_categorical('max_depth', [10,20,100]),
        'num_leaves' : trial.suggest_int('num_leaves', 1, 1000),
        'min_child_samples': trial.suggest_int('min_child_samples', 1, 300),
        'cat_smooth' : trial.suggest_int('min_data_per_groups', 1, 100)
    }
    model = LGBMRegressor(**param)  
    
    model.fit(train_x,train_y,eval_set=[(test_x,test_y)],early_stopping_rounds=100,verbose=False)
    
    preds = model.predict(test_x)
    
    rmse = mean_squared_error(test_y, preds,squared=False)
    
    return rmse

In [93]:
study = optuna.create_study(direction='minimize')
study.optimize(objective, n_trials=50)
print('Number of finished trials:', len(study.trials))
print('Best trial:', study.best_trial.params)

[I 2021-10-26 22:33:12,464] A new study created in memory with name: no-name-10816c20-3f08-4a1e-b92f-4afc9e3943f4
[W 2021-10-26 22:33:18,859] Trial 0 failed because of the following error: MemoryError((120822, 3365), dtype('float64'))
Traceback (most recent call last):
  File "C:\Users\andre\anaconda3\lib\site-packages\optuna\study\_optimize.py", line 213, in _run_trial
    value_or_values = func(trial)
  File "<ipython-input-92-367ce601768c>", line 24, in objective
    model.fit(train_x,train_y,eval_set=[(test_x,test_y)],early_stopping_rounds=100,verbose=False)
  File "C:\Users\andre\anaconda3\lib\site-packages\lightgbm\sklearn.py", line 822, in fit
    categorical_feature=categorical_feature, callbacks=callbacks, init_model=init_model)
  File "C:\Users\andre\anaconda3\lib\site-packages\lightgbm\sklearn.py", line 688, in fit
    callbacks=callbacks, init_model=init_model)
  File "C:\Users\andre\anaconda3\lib\site-packages\lightgbm\engine.py", line 228, in train
    booster = Booster(p

MemoryError: Unable to allocate 3.03 GiB for an array with shape (120822, 3365) and data type float64